### PR TITLE
[Sponsored by Cubepilot and Holybro] Cube ID via DroneCAN

### PR DESCRIPTION
**Pull request by julianoes**
This adds support for Cubepilot's Cube ID modules via DroneCAN, so that users are not limited to the Cube ID serial variant.

Thanks to [dirksavage88](https://github.com/dirksavage88), I borrowed quite a bit of code from https://github.com/PX4/PX4-Autopilot/pull/21647.

Tested using QGC master and DroneScanner.
~QGC v4.4.0 doesn't properly implement the "Fixed GPS GCS" setting (see https://github.com/mavlink/qgroundcontrol/issues/11673).~

Initial work for CubeID was sponsored by CubePilot.
Additionally, testing/fixup for the Holybro Remote ID was sponsored by Holybro.

### CODE REVIEW

**Comment by PetervdPerk-NXP** [comment]
Awesome, you managed to update the libuavcan and dsdl dependencies so that we can generate latest data types?

**Comment by julianoes** [comment]
@PetervdPerk-NXP I just updated it and nothing seemed to break, so I'm not sure I did it right...

**Comment by henrykotze** [comment]
We definitely need to test it on hardware, since I have updated previously to support newer data types, but there were issues translating the messages between FMU and Cannodes. I specify remember that the Servo Array Command didnt work as expected. 

I will see if I can find the Issue or some internal doc which describes the issue.

**Comment by julianoes** [comment]
@henrykotze or @AlexKlimaj any tips how I test whether the DroneCAN update breaks anything?

**Comment by julianoes** [comment]
@dagar what param name do you suggest, given this has subscriptions as well as publications? `UAVCAN_EN_RID`?

**Comment by julianoes** [comment]
This is working on my bench, I'd say this is on par with the serial implementation and we should be able to merge it once reviewed and flash issues are resolved.

**Comment by julianoes** [comment]
@dagar let me know what param name you want, and what else you need me to do to merge this.

**Comment by julianoes** [comment]
@henrykotze or @AlexKlimaj, please speak up how I have to test this, otherwise I'd like to get this in.

**Comment by AlexKlimaj** [comment]
I can test in a few days. @dakejahl may be able to test sooner.

**Comment by dakejahl** [comment]
I don't have the CAN version of the CubeID

**Comment by julianoes** [comment]
@dakejahl 

> I don't have the CAN version of the CubeID

That's what I have tested already.

However, I was told that other CAN devices such as CANnodes with servo array might be broken with the update.

**Comment by julianoes** [comment]
image

Updated and extended after testing with Holybro's Remote ID module as well which is running ArduRemoteID v1.14.

**Comment by julianoes** [comment]
FYI @BluemarkInnovations and @dirksavage88

**Comment by dirksavage88** [comment]
> FYI @BluemarkInnovations and @dirksavage88

@julianoes I can test the holybro, but I don't have the Cube ID CAN either

File: src/drivers/gnss/septentrio/septentrio.cpp [hunk 1]
```diff
@@ -140,7 +140,7 @@ constexpr const char *k_frequency_25_0hz = "msec40";
 constexpr const char *k_frequency_50_0hz = "msec20";
 
 px4::atomic<SeptentrioDriver *> SeptentrioDriver::_secondary_instance {nullptr};
-uint32_t SeptentrioDriver::k_supported_baud_rates[] {0, 38400, 57600, 115200, 230400, 460800, 500000, 576000, 921600, 1000000, 1500000};
+uint32_t SeptentrioDriver::k_supported_baud_rates[] {0, 38400, 57600, 115200, 230400, 460800, 921600};
```
Comment (dagar) [inline src/drivers/gnss/septentrio/septentrio.cpp:143]: Intentional?

**Comment by dagar** [comment]
This largely looks good, I'm just not sure about some of the message flow specifics given the range of different use cases (more convoluted than you might expect).

For example rather than just copying around `OPEN_DRONE_ID_ARM_STATUS` into uORB and then Mavlink streaming it out verbatim, shouldn't PX4 itself be generating `OPEN_DRONE_ID_ARM_STATUS` based on its overall ability to actually arm which INCLUDES consuming a remote id module's OPEN_DRONE_ID_ARM_STATUS?

**Comment by dagar** [comment]
Another case is `OPEN_DRONE_ID_SYSTEM`. It some cases it needs to be forwarded through from the GCS (with GPS) to the remote id module. In other cases it can be generated by PX4 itself using the takeoff location for operator_latitude/operator_longitude. It needs to be one or the other, but definitely not both right?

File: src/drivers/gnss/septentrio/septentrio.cpp (hunk 1 above)
Comment (julianoes) [reply src/drivers/gnss/septentrio/septentrio.cpp:143]: Pending https://github.com/PX4/PX4-Autopilot/pull/23415.

**Comment by julianoes** [comment]
Thanks for the review and comments:

> For example rather than just copying around OPEN_DRONE_ID_ARM_STATUS into uORB and then Mavlink streaming it out verbatim, shouldn't PX4 itself be generating OPEN_DRONE_ID_ARM_STATUS based on its overall ability to actually arm which INCLUDES consuming a remote id module's OPEN_DRONE_ID_ARM_STATUS?

The way it seems to work is that the remote ID module sends this message out and PX4 parses it, and sends it out to the GCS via MAVLink.

Are you saying PX4 should modify the message?

> Another case is OPEN_DRONE_ID_SYSTEM. It some cases it needs to be forwarded through from the GCS (with GPS) to the remote id module. In other cases it can be generated by PX4 itself using the takeoff location for operator_latitude/operator_longitude. It needs to be one or the other, but definitely not both right?

That's how I did it for the DroneCAN ones. If the GCS has sent a system message, then it is just forwarded. If not, then it will fall back to the takeoff location.

If you want we can add a param to explicitly configure that.

File: src/drivers/gnss/septentrio/septentrio.cpp (hunk 1 above)
Comment (julianoes) [reply src/drivers/gnss/septentrio/septentrio.cpp:143]: Dropped in the latest rebase.

**Comment by julianoes** [comment]
Rebased and force-pushed, removed the baudrate noise.

I suggest we merge this now, unless you have actionable review feedback.
Also, without test instructions or results regarding the DroneCAN update, there is nothing I can do but suggest to merge it and wait for the results to come in that way.

**Comment by AlexKlimaj** [comment]
Looks good to me. I just flashed on an ARK RTK GPS, validated that its working with old PX4. Then updated the ARKV6X with it and validated that its working.

```
nsh> uavcan status
Pool allocator status:
    Capacity hard/soft: 500/250 blocks
    Reserved:  26 blocks
    Allocated: 18 blocks

UAVCAN node status:
    Internal failures: 0
    Transfer errors:   198
    RX transfers:      4060
    TX transfers:      1700

CAN1 status:
    HW errors: 2889
    IO errors: 2889
    RX frames: 0
    TX frames: 2766
CAN2 status:
    HW errors: 0
    IO errors: 0
    RX frames: 19279
    TX frames: 2766

ESC outputs:
INFO  [mixer_module] Param prefix: UAVCAN_EC
control latency: 0 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
Channel Configuration:
Channel 0: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 1: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 2: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 3: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 4: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 5: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 6: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Channel 7: func:   0, value: 0, failsafe: 65535, disarmed: 65535, min: 1, max: 8191
Servo outputs:
INFO  [mixer_module] Param prefix: UAVCAN_SV
control latency: 0 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
Channel Configuration:
Channel 0: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 1: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 2: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 3: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 4: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 5: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 6: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000
Channel 7: func:   0, value: 500, failsafe: 500, disarmed: 500, min: 0, max: 1000

Sensor 'gnss':
name: uavcan_gnss
channel 0: node id 92 --> instance 0

Sensor 'gnss_relative':
name: uavcan_gnss_relative

Sensor 'mag':
name: uavcan_mag
channel 0: node id 92 --> instance 1

Online nodes (Node ID, Health, Mode):
     92 OK         OPERAT    

uavcan: cycle time: 50438 events, 1484783us elapsed, 29.44us avg, min 9us max 16212us 122.317us rms
uavcan: cycle interval: 50438 events, 2004.38us avg, min 15us max 16216us 1259.918us rms
nsh> 
```

**Comment by julianoes** [comment]
Great, thanks for testing @AlexKlimaj!

**Comment by hamishwillee** [comment]
Docs updates?

**Comment by julianoes** [comment]
Good point @hamishwillee. I'll plan that for tomorrow.

**Comment by julianoes** [comment]
docs update: https://github.com/PX4/PX4-user_guide/pull/3339

**Comment by julianoes** [comment]
Tried to make some flash space. (We really should update GCC at some point as that saves us some flash!)

**Comment by julianoes** [comment]
I do not understand why sometimes the mavlink include works and sometimes not, presumably a race during parallel compilation?

**Comment by julianoes** [comment]
> I do not understand why sometimes the mavlink include works and sometimes not, presumably a race during parallel compilation?

Now it passed, like it does locally for me.

I'm going to merge this now assuming it's fine given there were no more review comments.

**Comment by julianoes** [comment]
@dagar: regarding the last commit:

I did not manage to properly extract the open_drone_id MAVLink helpers into its own library. At build time there seemed to be a race condition between the MAVLink headers having been generated and the library requiring it.

Making the library depend on `mavlink_c_generate` seemed to work but made the IO boards fail which don't have MAVLink obviously.

Using ifs to avoid the open_drone_id files to be built for IO failed, as well as making it dependent on the MAVLink module. I assume that's because it is a circular dependency.

Instead I went with just moving the helpers into the MAVLink module and including the file in the UAVCAN driver as well.

Any other idea how I ought to do that?

**Comment by julianoes** [comment]
@mrpollo it would be good not to cancel all other builds when one fails. Otherwise, I have to fix one board at a time and then try again, just to find the next one that fails.

**Comment by hamishwillee** [comment]
@julianoes I've added the new uORB messages in https://github.com/PX4/PX4-user_guide/pull/3370

Are there any docs implications to https://docs.px4.io/main/en/peripherals/remote_id.html ?

**Comment by julianoes** [comment]
I think we already updated the docs @hamishwillee.

**Comment by AlexKlimaj** [comment]
I actually need sd bench in the ARV6X build. I use it on the production line to test the SD card.

**Comment by julianoes** [comment]
Oh, I'm sorry! Anything else you can remove instead?

Although, you could consider having a separate test image that gets reflashed with the production image once all tested.

**Comment by AlexKlimaj** [comment]
It looks like it all fits just adding it back in. I want to avoid having to flash multiple times in production. The time it takes really adds up.

**Comment by julianoes** [comment]
Ok good. Again, sorry about that but I find it super hard to get anything merged as any feature is going to hit the flash limit on at least some boards, and they all come with a slightly different set of modules and drivers installed, so it's impossible to know what's actually needed on each of them. The PR had been dragging for weeks, so I just took things in my own hands, disabled what I thought was "safe" and merged it. In an idea world I'd wait for everyone to give a thumbs up. Next time I know to ping you when I mess with an ARK board.

**Pull request merged**